File: prisma/refs/java/detect-squares.java
import java.util.*;

class DetectSquares {
    private final Map<Long, Integer> counts = new HashMap<>();
    private final List<int[]> points = new ArrayList<>();

    public DetectSquares() {}

    private long key(int x, int y) { return (long) x * 100_000 + y; }

    public void add(int[] point) {
        counts.merge(key(point[0], point[1]), 1, Integer::sum);
        points.add(point.clone());
    }

    public int count(int[] point) {
        int qx = point[0], qy = point[1], total = 0;
        for (int[] p : points) {
            int x = p[0], y = p[1];
            if (Math.abs(x - qx) != Math.abs(y - qy) || x == qx) continue;
            total += counts.getOrDefault(key(x, qy), 0) * counts.getOrDefault(key(qx, y), 0);
        }
        return total;
    }
}

File: prisma/refs/py/detect-squares.py
from collections import Counter


class DetectSquares:
    def __init__(self):
        self.counts = Counter()
        self.points = []

    def add(self, point):
        x, y = point
        self.counts[(x, y)] += 1
        self.points.append((x, y))

    def count(self, point):
        qx, qy = point
        total = 0
        for x, y in self.points:
            if abs(x - qx) != abs(y - qy) or x == qx:
                continue
            total += self.counts[(x, qy)] * self.counts[(qx, y)]
        return total
